### ISSUE TITLE
Managing playlists externally while cmus is running

### ISSUE THREAD

**Issue by sebw** [cmus/cmus#1066]
I want to rate my music with Ulauncher while cmus plays music, but I'm facing an issue.

1. I start cmus
2. I have 5 empty playlists (rated_1_stars, rated_2_stars, and so on)
3. I start playing music
4. In Ulauncher I type "rate 4" which triggers a script which adds the current song in the rated_4_stars playlist
5. I quit cmus
6. Playlists are emptied

Rating script is inspired from https://gist.github.com/Markus00000/ad8c0ff46290f9dc2887

Why are playlists reverted back to the start they were when cmust started. What am I doing wrong?

**Comment by pgaskin** [comment]
cmus loads the playlist on startup and saves it on exit. The proper way to do this would be with `cmus-remote`, but I don't think there's currently a command to add the current track to an arbitrary playlist without selecting it first.

**Comment by cubicvoid** [comment]
https://github.com/cmus/cmus/pull/1297 made progress on this recently but mentioned "design discussion" that needs to happen for the remaining work... would a PR adding a `pl-add-playing` command (to add the currently-playing song or album to a playlist) be welcome? If so, are there any constraints I should know before I try adding it?

**Comment by rodenet-hier** [comment]
> #1297 made progress on this recently but mentioned "design discussion" that needs to happen for the remaining work... would a PR adding a `pl-add-playing` command (to add the currently-playing song or album to a playlist) be welcome? If so, are there any constraints I should know before I try adding it?

That should be another low-hanging fruit in terms of design and implementation. I'd suggest building a set of commands around marking. `mark-playing` and `mark-pl <name>`/`pl-mark <name>` should be enough to implement `pl-add-playing` if combined with `win-add-p` for a start.

**Comment by cubicvoid** [comment]
> I'd suggest building a set of commands around marking. `mark-playing` and `mark-pl <name>/pl-mark <name>` should be enough to implement `pl-add-playing` if combined with `win-add-p` for a start.

`mark-playing`would interact confusingly with `win-add-p` since the music that's currently playing isn't necessarily markable in the current window. (Is there a clean way to resolve that? Even restricting to the sorted view isn't enough, since the currently playing song doesn't need to actually be in the library.)

What about adding an optional playlist name to `win-add-p`? It would be nice to bind a key to e.g. "win-add-p favorites" to add something to my favorites list without worrying about which playlist is currently selected. With that change, a `playing-add-p` command with similar semantics would fit in better. That also allows command-line actions with less concern about order-of-operations since the playlist wouldn't have to be marked in order to add things to it.

**Comment by rodenet-hier** [comment]
> `mark-playing`would interact confusingly with `win-add-p` since the music that's currently playing isn't necessarily markable in the current window. (Is there a clean way to resolve that? Even restricting to the sorted view isn't enough, since the currently playing song doesn't need to actually be in the library.)

You're right, I forgot the different views didn't share the marked tracks since I haven't used it all that much. The file path might have been an alternative but that would also bring us back to the design problem. 

> What about adding an optional playlist name to `win-add-p`? It would be nice to bind a key to e.g. "win-add-p favorites" to add something to my favorites list without worrying about which playlist is currently selected. With that change, a `playing-add-p` command with similar semantics would fit in better. That also allows command-line actions with less concern about order-of-operations since the playlist wouldn't have to be marked in order to add things to it.

The playlist marking should be easier as it's restricted to one view. So `win-add-p <playlist>` could be `mark-pl <playlist>` and `win-add-p` but neither seems useful for external management without marking commands. Your initial `playing-add-p <playlist>` / `pl-add-playing <playlist>` seems to be the most reasonable solution to me right now as well.

**Comment by cubicvoid** [comment]
> The playlist marking should be easier as it's restricted to one view. So `win-add-p <playlist>` could be `mark-pl <playlist>` and `win-add-p` but neither seems useful for external management without marking commands.

I'm not quite sure I understand which conclusion you're reaching on `win-add-p` specifically :-) I agree `pl-mark <playlist>` would be straightforward to implement (and I'm willing to if wanted) but I doubt I'd use it much. `win-add-p <playlist>` would then be equivalent to `pl-mark <playlist>` followed by `win-add-p`, but that wouldn't let you bind a single key to add marked/selected tracks to a target playlist the way explicit `win-add-p <playlist>` support would, so I'd prefer to add the optional playlist argument instead.

But maybe we agree and I'm just rephrasing what you said :-) To be clear, do you think a PR adding both `win-add-p <playlist>` and `playing-add-p <playlist>` commands (or maybe two PRs adding them separately) would be reasonable, or do you have reservations about the `win-add-p` change?
